modules/secrets/secrets.nix

```diff
@@ -73,8 +77,57 @@ in
          - Set `nix-bitcoin.secretsSetupMethod = "manual"` if you want to manually setup secrets
       '';
     };
+
+    generateSecretsScript = mkOption {
+      internal = true;
+      default = let
+        rpcauthSrc = builtins.fetchurl {
+          url = "https://raw.githubusercontent.com/bitcoin/bitcoin/d6cde007db9d3e6ee93bd98a9bbfdce9bfa9b15b/share/rpcauth/rpcauth.py";
+          sha256 = "189mpplam6yzizssrgiyv70c9899ggh8cac76j4n7v0xqzfip07n";
+        };
+        rpcauth = pkgs.writers.writeBash "rpcauth" ''
+          exec ${pkgs.python3}/bin/python ${rpcauthSrc} "$@"
+        '';
+      in pkgs.writers.writeBash "generate-secrets" ''
+        set -euo pipefail
+
+        export PATH=${lib.makeBinPath (with pkgs; [ coreutils gnugrep ])}
+
+        makePasswordSecret() {
+          # Passwords have alphabet {a-z, A-Z, 0-9} and ~119 bits of entropy
+          [[ -e $1 ]] || ${pkgs.pwgen}/bin/pwgen -s 20 1 > "$1"
+        }
+        makeBitcoinRPCPassword() {
+          user=$1
+          file=bitcoin-rpcpassword-$user
+          HMACfile=bitcoin-HMAC-$user
+          makePasswordSecret "$file"
+          if [[ $file -nt $HMACfile ]]; then
+            ${rpcauth} $user $(cat "$file") | grep rpcauth | cut -d ':' -f 2 > "$HMACfile"
+          fi
+        }
+        makeCert() {
+          name=$1
+          # Add leading comma if not empty
+          extraAltNames=''${2:+,}''${2:-}
+          if [[ ! -e $name-key || ! -e $name-cert ]]; then
+            ${pkgs.openssl}/bin/openssl req -x509 -newkey ec -pkeyopt ec_paramgen_curve:prime256v1 \
+              -sha256 -days 3650 -nodes -keyout "$name-key" -out "$name-cert" \
+              -subj "/CN=localhost/O=$name" \
+              -addext "subjectAltName=DNS:localhost,IP:127.0.0.1$extraAltNames"
```

Comment:
It's not worse than before but if a user changes lnd's rpcaddress, the cert would become invalid. Automatically regenerating the cert isn't a great solution either because the user will miss it and run into errors. Perhaps we should at least mention in the option description that changing this requires the cert to be regenerated?

Reply:
Good catch. Wouldn't it be better to just automatically regenerate the cert when details change? LND includes a config option for this [`tlsautorefresh`](https://github.com/lightningnetwork/lnd/blob/master/sample-lnd.conf#L53).